src/librustc_typeck/check/reconciliation.rs

```diff
@@ -0,0 +1,148 @@
+use crate::check::{FnCtxt, Needs, PlaceOp};
+use rustc_hir as hir;
+use rustc_middle::ty::adjustment::{Adjust, Adjustment, OverloadedDeref, PointerCast};
+use rustc_middle::ty::adjustment::{AllowTwoPhase, AutoBorrow, AutoBorrowMutability};
+use rustc_middle::ty::{self, Ty};
+
+impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    /// Convert auto-derefs, indices, etc of an expression from `Deref` and `Index`
```

Comment:
it'd be nice to expand this comment *just a bit* to explain that this is often done as a second pass because of the fact that we don't know whether the result is mutable or not until later